src/backend/distributed/utils/connection_cache.c

```diff
@@ -186,11 +182,17 @@ PurgeConnectionByKey(NodeConnectionKey *nodeConnectionKey)
 {
 	bool entryFound = false;
 	NodeConnectionEntry *nodeConnectionEntry = NULL;
+	PGconn *connection = NULL;
+
+	if (NodeConnectionHash != NULL)
+	{
+		nodeConnectionEntry = hash_search(NodeConnectionHash, nodeConnectionKey,
+										  HASH_REMOVE, &entryFound);
+	}
 
-	nodeConnectionEntry = hash_search(NodeConnectionHash, nodeConnectionKey, HASH_REMOVE,
-									  &entryFound);
 	if (entryFound)
```

Comment:
We could add PQsocket comparison here to make sure we're actually talking about the same connection.